internal/setup/setup.go

```diff
@@ -772,3 +817,66 @@ func Select(release *Release, slices []SliceKey) (*Selection, error) {
 
 	return selection, nil
 }
+
+// pathInfoToYAML converts a PathInfo object to a yamlPath object.
+func pathInfoToYAML(pi *PathInfo) (*yamlPath, error) {
+	path := &yamlPath{
+		Mode:    yamlMode(pi.Mode),
+		Mutable: pi.Mutable,
+		Until:   pi.Until,
+		Arch:    yamlArch{List: pi.Arch},
+	}
+	switch pi.Kind {
+	case DirPath:
+		path.Dir = true
+	case CopyPath:
+		path.Copy = pi.Info
+	case TextPath:
+		path.Text = &pi.Info
+	case SymlinkPath:
+		path.Symlink = pi.Info
+	case GlobPath:
+	default:
+		return nil, fmt.Errorf("internal error: unrecognised PathInfo type: %s", pi.Kind)
+	}
+	return path, nil
+}
+
+// sliceToYAML converts a Slice object to a yamlSlice object.
+func sliceToYAML(s *Slice) (*yamlSlice, error) {
+	slice := &yamlSlice{
+		Essential: make([]string, 0, len(s.Essential)),
+		Contents:  make(map[string]*yamlPath, len(s.Contents)),
+		Mutate:    s.Scripts.Mutate,
+	}
+	for _, key := range s.Essential {
+		slice.Essential = append(slice.Essential, key.String())
+	}
+	sort.Strings(slice.Essential)
+	for path, info := range s.Contents {
+		pi := info
```

Comment:
Are you intentfully copying the data to prevent mutating it? If so:

info := info

This makes your intention clear.

Why that said, it's weird that this is required here.  The signature of this function doesn't make it clear at all that it's about to mutate its parameter, and I can't see why it would.  If it does need to touch data, it should prevent mutation internally. If it doesn't mutate, then we don't need the line above here.

Reply:
Thanks for pointing it out.